File: PascalsTriangleII.py
class Solution:
    # @return a list of integers
    def getRow(self, rowIndex):
        """
        overlapped array
        """
        res = [[1 for x in xrange(rowIndex+1)] for y in range(2)]
        for i in xrange(1, rowIndex+1):
            for j in xrange(1, i):
                res[i%2][j] = res[(i-1)%2][j-1]+res[(i-1)%2][j]
        return res[rowIndex%2]

if __name__ == '__main__':
    so = Solution()
    print so.getRow(1)

File: CPP/Second/pascals_triangle_ii.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    // careful, should be right to left
    // why?
    // assume 1 2 1 => 1 3 3 1
    // if from left to right, the second 3 becomes 4
    // add from right to left could get rid of these
    vector<int> getRow(int rowIndex) {
        vector<int> res(rowIndex + 1, 0);
        res[0] = 1;
        for (int i = 1; i <= rowIndex; i++) {
            for (int j = i - 1; j > 0; j--) {
                res[j] = res[j-1] + res[j];
            }
            res[i] = 1;
        }
        return res;
    }
};

int main(int argc, char *argv[])
{
    Solution so;
    so.getRow(5);
    return 0;
}
